SharedUtils/src/main/java/com/braintreepayments/api/SignatureVerification.java

```diff
@@ -47,38 +43,28 @@ static boolean isSignatureValid(Context context, String packageName,
             return false;
         }
 
-        InputStream certStream = null;
-        boolean validated = (signatures.length != 0);
-        for (Signature signature : signatures) {
-            try {
-                certStream = new ByteArrayInputStream(signature.toByteArray());
-
-                X509Certificate x509Cert =
-                        (X509Certificate) CertificateFactory.getInstance("X509")
-                                .generateCertificate(certStream);
-
-                String subject = x509Cert.getSubjectX500Principal().getName();
-                String issuer = x509Cert.getIssuerX500Principal().getName();
-                int actualPublicKeyHashCode = x509Cert.getPublicKey().hashCode();
+        if (signatures.length == 0) {
+            return false;
+        }
 
-                validated &= (certificateSubject.equals(subject) &&
-                        certificateIssuer.equals(issuer) &&
-                        publicKeyHashCode == actualPublicKeyHashCode);
+        for (Signature signature : signatures) {
+            if (signature != null) {
+                String currentSignature;
+                try {
+                    MessageDigest md = MessageDigest.getInstance("SHA-256");
+                    md.update(signature.toByteArray());
+                    currentSignature = Base64.encodeToString(md.digest(), Base64.DEFAULT);
+                } catch (NoSuchAlgorithmException e) {
```

Comment:
Will this prevent us from having issues like this in the future? Just to summarize - the problem was that the hashCode() functionality changed b/w Android 11 and 12, and this issue wasn't specific to Venmo?

Reply:
Yes there were no guarantees about the underlying implementation of the `hashCode` function, but this new approach relies on the SHA-256 algorithm and base 64 encoding. I'm not a cryptography expert, but as far as I know the algorithm should always produce the same output for a given input.